File: bls.py
import numpy as np
import pandas as pd

def bls_search(t, y, periods, n_bins=200, d_min=0.01, d_max=0.1):
    """Conducts a Box Least Squares search to find power across a
    provided grid.

    Args:
        t (np.array): Array of the times of observation
        y (np.array): Array of flattened and zero-based observations
        periods (np.array): Array of periods to check
        n_bins (int): The number of bins to divide the folded signal into
        d_min (float): The minimum size of a duration, in normalized phase units
        d_max (float): The maximum size of a duration, in normalized phase units

    Returns
        (pd.DataFrame): A dataframe containing best values for each period
    """

    N = len(y)
    results = []
   
    # For binning to work well, we constrain our starting and ending durations into
    # multiples of bins. This might not work out EXACTLY to the desired start and
    # stops, but it is close enough and provides big speed gains
    L_min = max(1, int(d_min * n_bins))
    L_max = max(L_min + 1, int(d_max * n_bins))
   
    # Looping over all the periods
    for P in periods:
        # Binning & Cumulative Summing for increased speed
        phase = (t % P) / P
        bin_idx = np.clip((phase * n_bins).astype(int), 0, n_bins - 1)
        s_bins = np.bincount(bin_idx, weights=y, minlength=n_bins)
        r_bins = np.bincount(bin_idx, minlength=n_bins)
        
        S = np.cumsum(np.insert(np.tile(s_bins, 2), 0, 0))
        R = np.cumsum(np.insert(np.tile(r_bins, 2), 0, 0))
       
        best_pwr_for_P = -1
        best_metadata = {}

        # Looping over all the possible durations
        for L in range(L_min, L_max + 1):
            # We vectorize to compute all the possible box positions simultaneously
            i = np.arange(n_bins)
            j = i + L
            s_sum = S[j] - S[i]
            r_sum = R[j] - R[i]
            
            # Calculate Power, tracking only dips
            mask = (r_sum > 0) & (r_sum < N) & (s_sum < 0)
            pwr = np.zeros(n_bins)
            pwr[mask] = (s_sum[mask]**2) / (r_sum[mask] * (N - r_sum[mask]))
            
            # Find the best start position 'i' for THIS specific L
            current_max_idx = np.argmax(pwr)
            current_max_pwr = pwr[current_max_idx]
            
            if current_max_pwr > best_pwr_for_P:
                best_pwr_for_P = current_max_pwr
                # Store the raw ingredients
                best_metadata = {
                    'period': P,
                    'power': best_pwr_for_P,
                    'duration_bins': L,
                    'start_bin': current_max_idx,
                    's': s_sum[current_max_idx],
                    'r': r_sum[current_max_idx]
                }
        
        results.append(best_metadata)
        
    return pd.DataFrame(results)

File: bls.r
library(tidyverse)

bls_search <- function(t, y, periods, n_bins = 200, d_min = 0.01, d_max = 0.1) {
  N <- length(y)
  num_periods <- length(periods)
  
  # Pre-calculate bin counts for duration L
  L_min <- max(1, floor(d_min * n_bins))
  L_max <- max(L_min + 1, floor(d_max * n_bins))
  L_range <- L_min:L_max
  
  # Pre-allocate vectors for speed
  best_powers <- numeric(num_periods)
  best_L      <- integer(num_periods)
  best_start  <- integer(num_periods)
  best_s      <- numeric(num_periods)
  best_r      <- numeric(num_periods)
 
  # Loop over periods
  for (p_idx in seq_along(periods)) {
    P <- periods[p_idx]
    
    # Binning
    phase <- (t %% P) / P
    bin_idx <- floor(phase * n_bins) + 1
    bin_idx[bin_idx > n_bins] <- n_bins
    
    # Counts (r) and Weighted Sums (s)
    r_bins <- tabulate(bin_idx, nbins = n_bins)
    s_bins <- as.numeric(tapply(y, bin_idx, sum))
    s_bins[is.na(s_bins)] <- 0
    
    # 2. Tiling & Cumulative Sums
    S <- cumsum(c(0, s_bins, s_bins))
    R <- cumsum(c(0, r_bins, r_bins))
    
    # Initialize trackers for this specific Period
    max_pwr_P <- -1
    
    # Loop over durations
    for (L in L_range) {
      i_idx <- 1:n_bins # We vectorize the position, to compute all at once
      j_idx <- i_idx + L
      
      s_sum <- S[j_idx + 1] - S[i_idx]
      r_sum <- R[j_idx + 1] - R[i_idx]
      
      # Calculate Power, ensuring only dips counting
      mask <- (r_sum > 0) & (r_sum < N) & (s_sum < 0)
      
      if(any(mask)) {
        pwr <- rep(0, n_bins)
        pwr[mask] <- (s_sum[mask]^2) / (r_sum[mask] * (N - r_sum[mask]))
        
        # Identify the best start position for this L
        loc_max_idx <- which.max(pwr)
        loc_max_pwr <- pwr[loc_max_idx]
        
        # If this is the best L we've seen for this Period, save the metadata
        if (loc_max_pwr > max_pwr_P) {
          max_pwr_P <- loc_max_pwr
          best_powers[p_idx] <- loc_max_pwr
          best_L[p_idx]      <- L
          best_start[p_idx]  <- loc_max_idx
          best_s[p_idx]      <- s_sum[loc_max_idx]
          best_r[p_idx]      <- r_sum[loc_max_idx]
        }
      }
    }
  }
  
  # 4. Return the Final Results Table
  return(tibble(
    period = periods,
    power = best_powers,
    duration_bins = best_L,
    start_bin = best_start,
    s = best_s,
    r = best_r
  ))
}
